Sort SkinSection in alphabetical order

rednir opened the pull request:
Excluding `<Random Skin>` and the built-in skins, all skins are sorted alphabetically as opposed to by date added. This matches behavior from stable, and it is much easier to find skins when you have a large amount.
Also, `<Random Skin>` makes more sense to be at the very top of the list imo.

This is my first PR so I realize I may have gone about doing this in the worst way possible.

rednir committed:
Sort SkinSection in alphabetical order

rednir committed:
Merge branch 'master' into sort-skins-alphabetically

rednir committed:
Merge branch 'master' of https://github.com/ppy/osu into sort-skins-alphabetically

rednir committed:
Merge branch 'sort-skins-alphabetically' of https://github.com/rednir/osu into sort-skins-alphabetically

rednir committed:
Remove whitespace

bdach commented on osu.Game/Overlays/Settings/Sections/SkinSection.cs:
```diff
@@ -112,21 +112,19 @@ private void load(OsuConfigManager config)
         private void itemUpdated(ValueChangedEvent<WeakReference<SkinInfo>> weakItem)
         {
             if (weakItem.NewValue.TryGetTarget(out var item))
-                Schedule(() => skinDropdown.Items = skinDropdown.Items.Where(i => !i.Equals(item)).Append(item).ToArray());
```
The `.Where(i => !i.Equals(item))` check has been incorrectly removed. `itemUpdated` is also fired on skin updates, not necessarily always additions.

bdach commented on osu.Game/Overlays/Settings/Sections/SkinSection.cs:
```diff
@@ -135,6 +133,24 @@ private void itemRemoved(ValueChangedEvent<WeakReference<SkinInfo>> weakItem)
 
+        private List<SkinInfo> sortList(List<SkinInfo> skinsList)
+        {
+            skinsList.Sort((a, b) => String.Compare(a.Name, b.Name, StringComparison.Ordinal));
+            for (int i = 0; i < skinsList.Count; i++)
+            {
+                // insert lazer built-in skins before user skins
+                if (skinsList[i].ID <= 0) {
+                    var itemToMove = skinsList[i];
+                    skinsList.RemoveAt(i);
+                    skinsList.Insert(0, itemToMove);
+                }
+            }
+            skinsList.RemoveAll(s => s.ID == SkinInfo.RANDOM_SKIN);
+            skinsList.Insert(0, random_skin_info);
```
Several issues with this:

* This needlessly removes three items and reinserts them back to the start again, incurring unnecessary six linear operations on the entire list, while it should only be sorting the section of the list that this cares about (so user skins). `List<T>.Sort()` has an overload that takes an index and count.
* Due to the above this also wrongly pushes the `Random Skin` option to the top, while it intentionally used to be third in the order.

bdach reviewed (changes requested):
In addition to these comments, please see CI errors due to whitespace that does not adhere to our code convention.

rednir committed:
fixed itemUpdated()

rednir committed:
Only user skins are sorted

rednir committed:
Merge branch 'sort-skins-alphabetically' of https://github.com/rednir/osu into sort-skins-alphabetically

rednir commented:
Didn't use the `List<T>.Sort()` overload you mentioned because it required a parameter that kept returning a "user locale could change string comparison" error, so instead copied user skins to a new list and then sorted that. Should that be fine?

rednir committed:
Remove another whitespace

rednir committed:
Remove another whitespace

rednir committed:
Merge branch 'sort-skins-alphabetically' of https://github.com/rednir/osu into sort-skins-alphabetically

rednir committed:
Merge branch 'master' into sort-skins-alphabetically

rednir committed:
Satisfy AppVeyor

rednir committed:
Satisfy AppVeyor

rednir committed:
Merge branch 'sort-skins-alphabetically' of https://github.com/rednir/osu into sort-skins-alphabetically

rednir committed:
Fix firstNonDefault staying as -1

bdach committed:
Make first non-default skin index a property

The previous code was very brittle - it was not always updating
properly, and seems to have worked either by a carefully crafted set of
circumstances, or just plain coincidence.

Having this be a get-only property avoids potential error in the future
caused by not updating the index properly, at the expense of an added
linear lookup.

rednir commented:
Seems like I still can't use git properly. Should I do something about this commit spam?

bdach commented:
No need. I'm in the middle of applying changes to this branch anyways.

bdach committed:
Refactor skin sorting method

* Rename to `sortUserSkins` to convey meaning better.
* Sort in-place instead of slicing the list.
* Change to `void` to avoid misleading users that the method returns
  a new list instance.
* Fix typo in comment.

bdach committed:
Schedule entire operation for safety

Also removes a redundant list copy.

bdach commented:
I've made a few changes to this. I think I'm personally OK with how this looks now, but would appreciate a double-check on this to be sure.

@rednir make sure to read through the changes and the commit messages to see what was changed and why.

rednir commented:
I really appreciate the comments and your support, I'll definitely keep in mind these practices if I ever open another PR. Thanks!

smoogipoo committed:
Merge branch 'master' into sort-skins-alphabetically

fem-al-e commented:
While renaming my skin collection today, I noticed that _lowercase letters are sorted separately_ from their uppercase variants, appearing at the bottom of SkinSection.

I couldn't find this being mentioned elsewhere, and I doubt that this is intended behavior, so I'm hoping that it's fine for me to bring it up here.

https://github.com/user-attachments/assets/041c394e-a756-46ac-9fe9-db5644b9540e

The list sorting order in the video above (excluding the default options) seems to be roughly:
1. Non-Alpha characters (here: dash)
2. Uppercase letters (A-Z)
3. More non-alpha characters (here: square brackets)
4. Lowercase letters (a-z)

I did not check for other non-alpha characters (numbers, curly brackets, parantheses etc.) yet.
